File: src/main/java/com/paragon/client/systems/ui/window/impl/Window.java
package com.paragon.client.systems.ui.window.impl;

import com.paragon.api.util.render.TextRenderer;

public abstract class Window implements TextRenderer {

    private String title;

    private float x;
    private float y;
    private float width;
    private float height;

    private float lastX;
    private float lastY;

    public Window(String title, float x, float y, float width, float height) {
        this.title = title;
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }

    public abstract void drawWindow(int mouseX, int mouseY);
    public abstract void mouseClicked(int mouseX, int mouseY, int mouseButton);
    public abstract void mouseReleased(int mouseX, int mouseY, int mouseButton);
    public abstract void keyTyped(char typedChar, int keyCode);

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public float getX() {
        return x;
    }

    public void setX(float x) {
        this.x = x;
    }

    public float getY() {
        return y;
    }

    public void setY(float y) {
        this.y = y;
    }

    public float getWidth() {
        return width;
    }

    public void setWidth(float width) {
        this.width = width;
    }

    public float getHeight() {
        return height;
    }

    public void setHeight(float height) {
        this.height = height;
    }

    public float getLastX() {
        return lastX;
    }

    public void setLastX(float lastX) {
        this.lastX = lastX;
    }

    public float getLastY() {
        return lastY;
    }

    public void setLastY(float lastY) {
        this.lastY = lastY;
    }
}

File: src/main/java/com/paragon/client/ui/configuration/window/window/Window.kt
package com.paragon.client.ui.configuration.window.window

import com.paragon.api.util.render.ITextRenderer
import com.paragon.client.ui.util.Click

/**
 * @author Wolfsurge
 */
abstract class Window(var x: Float, var y: Float, var width: Float, var height: Float) : ITextRenderer {

    var scroll = 0f
    var scrollFactor = 0f

    abstract fun draw(mouseX: Int, mouseY: Int)
    abstract fun mouseClicked(mouseX: Int, mouseY: Int, button: Click)
    abstract fun mouseReleased(mouseX: Int, mouseY: Int, button: Click)
    abstract fun keyTyped(character: Char, keyCode: Int)

}